autogen/agentchat/contrib/gpt_assistant_agent.py

```diff
@@ -158,68 +162,113 @@ def _invoke_assistant(
+        if run.status == "failed":
+            logger.error(f"Run: {run.id} Thread: {assistant_thread.id}: failed...")
+            if run.last_error:
+                response = {
+                    "role": "assistant",
+                    "content": str(run.last_error),
+                }
+            else:
+                response = {
+                    "role": "assistant",
+                    "content": "Failed",
+                }
+            return response
+        elif run.status == "expired":
+            logger.warn(f"Run: {run.id} Thread: {assistant_thread.id}: expired...")
+            response = {
+                "role": "assistant",
+                "content": "Expired",
+            }
+            return response
+        elif run.status == "cancelled":
+            logger.warn(f"Run: {run.id} Thread: {assistant_thread.id}: cancelled...")
+            response = {
+                "role": "assistant",
+                "content": "Cancelled",
+            }
+            return response
+        elif run.status == "completed":
+            logger.info(f"Run: {run.id} Thread: {assistant_thread.id}: completed...")
+            response_messages = self._openai_client.beta.threads.messages.list(assistant_thread.id, order="asc")
+            new_messages = []
+            for msg in response_messages:
+                if msg.run_id == run.id:
+                    for content in msg.content:
+                        if content.type == "text":
+                            new_messages.append(
+                                {"role": msg.role, "content": self._format_assistant_message(content.text)}
+                            )
+                        elif content.type == "image_file":
+                            new_messages.append(
+                                {
+                                    "role": msg.role,
+                                    "content": f"Recieved file id={content.image_file.file_id}",
+                                }
+                            )
+            response = {
+                "role": new_messages[-1]["role"],
+                "content": "",
+            }
+            for message in new_messages:
+                # just logging or do something with the intermediate messages?
+                # if current response is not empty and there is more, append new lines
+                if len(response["content"]) > 0:
+                    response["content"] += "\n\n"
+                response["content"] += message["content"]
+            return response
+
+    def _get_run_response(self, assistant_thread, run):
         """
         Waits for and processes the response of a run from the OpenAI assistant.
-
         Args:
+            assistant_thread: The thread object for the assistant.
             run: The run object initiated with the OpenAI assistant.
-
-        Returns:
-            Updated run object, status of the run, and response messages.
         """
         while True:
```

Comment:
Is it better to check `_cancel_run()` here?

Reply:
Is the suggestion to remove current line 288-289. And instead put them right after 244. If so, that makes sense to me. Though I think we don't yet have a method to allow the user to set the value of `self.cancellation_requested`. If I understood the issue correctly, shall we also add that?

Reply:
Yes, I am with you on that

Reply:
It can move after submission but I figure to let it finish submitting an action before trying to cancel it. Looks good otherwise.